### PR TITLE
Added --client-addr option

### PR DESCRIPTION
**Pull request by nigelhorne**
With this only remote connections from the given IP address will be accepted.  My use case is an SMTP forward I'm using and I don't want it to be used as an open relay and I don't have the option of adding a firewall.

### CODE REVIEW

File: src/tcptunnel.c [hunk 1]
```diff
@@ -280,6 +280,7 @@ int wait_for_clients(void)
 	if(settings.client_addr && (strcmp(inet_ntoa(rc.client_addr.sin_addr), options.client_addr) != 0)) {
 		if (settings.log)
 			printf("> %s tcptunnel: refused request request from %s\n", get_current_timestamp(), inet_ntoa(rc.client_addr.sin_addr));
+		close(rc.client.socket);
```
Comment (vakuum) [inline src/tcptunnel.c]: Does this fix the issue in #5?

Comment (nigelhorne) [inline src/tcptunnel.c]: I do not believe so, unless the link is being bombarded by scanners, however it is not impossible.

File: src/tcptunnel.c [hunk 2]
```diff
@@ -144,6 +145,11 @@ void set_options(int argc, char *argv[])
 				break;
 			}
 
+			case CLIENT_ADDR_OPTION:
+				options.client_addr = optarg;
+				settings.client_addr = 1;
+				break;
+
```
Comment (vakuum) [inline src/tcptunnel.c]: Can you add braces around the case block? So it matches the used code style.

```c
case CLIENT_ADDRESS_OPTION:
{
	options.client_addr = optarg;
	settings.client_addr = 1;
	break;
}
```

File: src/tcptunnel.c [hunk 3]
```diff
@@ -41,14 +41,15 @@
...
-struct struct_settings settings = { 0, 0, 0, 0, 0, 0, 0, 0 };
+struct struct_settings settings = { 0, 0, 0, 0, 0, 0, 0, 0, 0 };
 
 static struct option long_options[] = {
 	{ "local-port",    required_argument, NULL, LOCAL_PORT_OPTION },
 	{ "remote-host",   required_argument, NULL, REMOTE_HOST_OPTION },
 	{ "remote-port",   required_argument, NULL, REMOTE_PORT_OPTION },
 	{ "bind-address",  required_argument, NULL, BIND_ADDRESS_OPTION },
 	{ "buffer-size",   required_argument, NULL, BUFFER_SIZE_OPTION },
+	{ "client-addr",   required_argument, NULL, CLIENT_ADDR_OPTION },
```
Comment (vakuum) [inline src/tcptunnel.c]: Can you rename `CLIENT_ADDR_OPTION` into `CLIENT_ADDRESS_OPTION`? So it matches the naming of `BIND_ADDRESS_OPTION`.

File: src/tcptunnel.c [hunk 4]
```diff
@@ -271,6 +277,13 @@ int wait_for_clients(void)
 		return 1;
 	}
 
+	if(settings.client_addr && (strcmp(inet_ntoa(rc.client_addr.sin_addr), options.client_addr) != 0)) {
+		if (settings.log)
+			printf("> %s tcptunnel: refused request request from %s\n", get_current_timestamp(), inet_ntoa(rc.client_addr.sin_addr));
```
Comment (vakuum) [inline src/tcptunnel.c]: Can you change "refused request request" into "refused request".

File: src/tcptunnel.c [hunk 5]
```diff
@@ -271,6 +277,13 @@ int wait_for_clients(void)
...
+		close(rc.client_socket);
+		return 1;
+	}
```
Comment (vakuum) [inline src/tcptunnel.c]: Can you reformat this block? So it matches the used code style.

```c
if (settings.client_addr && (strcmp(inet_ntoa(rc.client_addr.sin_addr), options.client_addr) != 0))
{
	if (settings.log)
	{
		printf("> %s tcptunnel: refused request from %s\n", get_current_timestamp(), inet_ntoa(rc.client_addr.sin_addr));
	}
	close(rc.client_socket);
	return 1;
}
```

File: src/tcptunnel.c [hunk 6]
```diff
@@ -460,6 +473,7 @@ Options:\n\
   --remote-port=PORT   remote port\n\
   --remote-host=HOST   remote host\n\
   --bind-address=IP    bind address\n\
+  --client-addr=IP     only accept connections from this address\n\
```
Comment (vakuum) [inline src/tcptunnel.c]: As above `--client-address`.

File: src/tcptunnel.h [hunk 7]
```diff
@@ -13,6 +13,7 @@
 #define STAY_ALIVE_OPTION   'h'
 #define HELP_OPTION         'i'
 #define VERSION_OPTION      'j'
+#define	CLIENT_ADDR_OPTION  'k'
```
Comment (vakuum) [inline src/tcptunnel.h]: Can you replace the tab character with a space?

```c
#define CLIENT_ADDRESS_OPTION ...
```

File: src/tcptunnel.h [hunk 8]
```diff
@@ -46,13 +47,15 @@ struct struct_settings {
...
 	unsigned int stay_alive   : 1;
+	unsigned int client_addr  : 1;
 };
 
 struct struct_options {
 	char *local_port;
 	char *remote_host;
 	char *remote_port;
 	char *bind_address;
+	const char *client_addr;
```
Comment (vakuum) [inline src/tcptunnel.h]: As above `client_address`.

File: src/tcptunnel.h [hunk 9]
```diff
@@ -46,13 +47,15 @@ struct struct_settings {
 	unsigned int fork         : 1;
 	unsigned int log          : 1;
 	unsigned int stay_alive   : 1;
+	unsigned int client_addr  : 1;
```
Comment (vakuum) [inline src/tcptunnel.h]: As above `client_address`.

File: src/tcptunnel.h (hunk 7 above)
Comment (vakuum) [inline src/tcptunnel.h]: Can you move the `CLIENT_ADDRESS_OPTION` line under the `BIND_ADDRESS_OPTION` line? So it matches the options order.

**Comment by nigelhorne** [comment]
Thanks!

**Pull request merged**